File: frontend/js/charger_code_tables.js
// Code-letter mappings for charger traceability serial numbers, per the
// "AEIDTH Traceability Detail" sheet (F.No. ADT_R&D_F003_V1.0).
//
// Mirrors the server's bms/charger_code_tables.py (aeidthocpp repo) so the
// Lot page's dropdowns render instantly instead of waiting on a network
// round trip to /api/charger-lot-options. These are fixed, rarely-changing
// choice lists (not per-lot data), so a local copy is safe - but it DOES
// mean the two copies can drift if the server table is ever edited without
// updating this file too. Lot *creation* still goes to the server (it needs
// an atomic, server-side sequence counter) - only the dropdown contents are
// mirrored here.
(function () {
  const VOLTAGE_AMP_CHOICES = [
    ["5825", "58V25A-5825"],
    ["7325", "73V25A-7325"],
  ];

  const VARIANT_CHOICES = [
    ["A", "CAN+IP-A"],
    ["B", "NON CAN+IP-B"],
  ];

  const CONNECTOR_CHOICES = [
    ["A", "Anderson SB75-A"],
    ["B", "Anderson SB50-B"],
    ["C", "Chagori-C"],
  ];

  const MS_ID_CHOICES = [
    ["A", "SHIGAN-A"],
    ["B", "SBT-B"],
    ["C", "OSRIM-C"],
    ["D", "TADASHI-D"],
    ["E", "IEMS-E"],
    ["G", "IKIO-G"],
  ];

  const MONTH_CHOICES = [
    ["A", "Jan"], ["B", "Feb"], ["C", "Mar"], ["D", "Apr"],
    ["E", "May"], ["F", "Jun"], ["G", "Jul"], ["H", "Aug"],
    ["I", "Sep"], ["J", "Oct"], ["K", "Nov"], ["L", "Dec"],
  ];
  const MONTH_CODES = MONTH_CHOICES.map(([c]) => c);

  const YEAR_CODE_BASE_YEAR = 2025; // "A" -> 2025, "B" -> 2026, ...

  const DEFAULT_VOLTAGE_AMP_CODE = "5825";
  const DEFAULT_VARIANT_CODE = "B";   // NON CAN+IP-B
  const DEFAULT_CONNECTOR_CODE = "A"; // Anderson SB75-A
  const DEFAULT_MS_ID_CODE = "G";     // IKIO-G

  function monthCodeFor(date) {
    return MONTH_CODES[date.getMonth()]; // 0-11 -> 'A'..'L'
  }

  function yearCodeFor(date) {
    const offset = date.getFullYear() - YEAR_CODE_BASE_YEAR;
    if (offset < 0 || offset > 25) {
      throw new Error(`Year ${date.getFullYear()} is out of the supported traceability code range`);
    }
    return String.fromCharCode("A".charCodeAt(0) + offset);
  }

  function toChoiceList(pairs) {
    return pairs.map(([code, label]) => ({ code, label }));
  }

  // Same shape as the server's charger_code_tables.options_payload(), so
  // lot.js can consume it identically regardless of source.
  function optionsPayload(now) {
    const date = now || new Date();
    return {
      voltage_amp: toChoiceList(VOLTAGE_AMP_CHOICES),
      variant: toChoiceList(VARIANT_CHOICES),
      connector: toChoiceList(CONNECTOR_CHOICES),
      ms_id: toChoiceList(MS_ID_CHOICES),
      month: toChoiceList(MONTH_CHOICES),
      year: [0, 1, 2, 3].map((i) => ({
        code: String.fromCharCode("A".charCodeAt(0) + i),
        label: String(YEAR_CODE_BASE_YEAR + i),
      })),
      defaults: {
        voltage_amp_code: DEFAULT_VOLTAGE_AMP_CODE,
        variant_code: DEFAULT_VARIANT_CODE,
        connector_code: DEFAULT_CONNECTOR_CODE,
        ms_id_code: DEFAULT_MS_ID_CODE,
        month_code: monthCodeFor(date),
        year_code: yearCodeFor(date),
      },
    };
  }

  window.ChargerCodeTables = { optionsPayload };
})();

File: backend/core/charger_code_tables.py
"""Code-letter mappings for charger traceability serial numbers, per the
"AEIDTH Traceability Detail" sheet (F.No. ADT_R&D_F003_V1.0).

Mirrors the server's bms/charger_code_tables.py (aeidthocpp repo) - kept in
sync manually, same as frontend/js/charger_code_tables.js already does for
the Lot page's dropdowns. Used by backend/tools/gen_serial_qr_sheet.py so it
can build a lot's prefix locally (offline, no server round-trip) using the
exact same validation/format rules the server itself enforces.
"""

from typing import Optional

# (code, label) pairs - code is what goes into the serial number, label is
# what's shown in the UI dropdown.
VOLTAGE_AMP_CHOICES = [
    ("5825", "58V25A-5825"),
    ("7325", "73V25A-7325"),
]

VARIANT_CHOICES = [
    ("A", "CAN+IP-A"),
    ("B", "NON CAN+IP-B"),
]

CONNECTOR_CHOICES = [
    ("A", "Anderson SB75-A"),
    ("B", "Anderson SB50-B"),
    ("C", "Chagori-C"),
]

MS_ID_CHOICES = [
    ("A", "SHIGAN-A"),
    ("B", "SBT-B"),
    ("C", "OSRIM-C"),
    ("D", "TADASHI-D"),
    ("E", "IEMS-E"),
    ("G", "IKIO-G"),
]

MONTH_CHOICES = [
    ("A", "Jan"), ("B", "Feb"), ("C", "Mar"), ("D", "Apr"),
    ("E", "May"), ("F", "Jun"), ("G", "Jul"), ("H", "Aug"),
    ("I", "Sep"), ("J", "Oct"), ("K", "Nov"), ("L", "Dec"),
]
MONTH_CODES = [c for c, _ in MONTH_CHOICES]

YEAR_CODE_BASE_YEAR = 2025  # "A" -> 2025, "B" -> 2026, "C" -> 2027, "D" -> 2028, ...

# Defaults per the traceability sheet / the requested default part config.
DEFAULT_VOLTAGE_AMP_CODE = "5825"
DEFAULT_VARIANT_CODE = "B"       # NON CAN+IP-B
DEFAULT_CONNECTOR_CODE = "A"     # Anderson SB75-A
DEFAULT_MS_ID_CODE = "G"         # IKIO-G

PRODUCT_CODE = "CC"  # Charger


def month_code_for(dt) -> str:
    """dt.month is 1-12 -> 'A'..'L'."""
    return MONTH_CODES[dt.month - 1]


def year_code_for(dt) -> str:
    """2025 -> 'A', 2026 -> 'B', ... Grows past 'Z' is not handled - not a
    realistic concern for this product's lifetime, but raises clearly if hit."""
    offset = dt.year - YEAR_CODE_BASE_YEAR
    if offset < 0 or offset > 25:
        raise ValueError(f"Year {dt.year} is out of the supported traceability code range")
    return chr(ord("A") + offset)


def build_prefix(voltage_amp_code, variant_code, connector_code, ms_id_code, month_code, year_code) -> str:
    return f"{PRODUCT_CODE}{voltage_amp_code}{variant_code}{connector_code}{ms_id_code}{month_code}{year_code}"


def _valid_codes(choices):
    return {c for c, _ in choices}


def validate_codes(voltage_amp_code, variant_code, connector_code, ms_id_code, month_code, year_code) -> Optional[str]:
    """Returns an error message if any code is invalid, else None."""
    if voltage_amp_code not in _valid_codes(VOLTAGE_AMP_CHOICES):
        return f"Invalid voltage_amp_code: {voltage_amp_code}"
    if variant_code not in _valid_codes(VARIANT_CHOICES):
        return f"Invalid variant_code: {variant_code}"
    if connector_code not in _valid_codes(CONNECTOR_CHOICES):
        return f"Invalid connector_code: {connector_code}"
    if ms_id_code not in _valid_codes(MS_ID_CHOICES):
        return f"Invalid ms_id_code: {ms_id_code}"
    if month_code not in MONTH_CODES:
        return f"Invalid month_code: {month_code}"
    if not (isinstance(year_code, str) and len(year_code) == 1 and "A" <= year_code <= "Z"):
        return f"Invalid year_code: {year_code}"
    return None
